templates/todo/common/auth/js/azureDeveloperCliCredential.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+    AccessToken,
+    TokenCredential,
+} from "@azure/core-auth";
+import { CredentialUnavailableError } from "@azure/identity";
+import child_process from "child_process";
+
+export class AzureDeveloperCliCredential implements TokenCredential {
+    public async getToken(
+        scopes: string | string[],
+    ): Promise<AccessToken> {
+        if (typeof(scopes) === "string") {
+            scopes = [scopes];
+        }
+
+        try {
+            const {stdout, error} = await getAccessTokenFromAzd(scopes);
+            if (error) {
+                throw new CredentialUnavailableError(`failed to call azd get-access-token: ${error}`);    
+            }
+
+            const resp: { token: string, expiresOn: string} = JSON.parse(stdout);
+
+            return {
+                token: resp.token,
+                expiresOnTimestamp: new Date(resp.expiresOn).getTime()
+            };
+        } catch (err: unknown) {
+            throw new CredentialUnavailableError(`failed to call azd get-access-token: ${err}`);   
```

Comment:
is the `CredentialUnavailableError` used by the ChainedCredential to move on to the next credential?.

If yes, should we return other error when we are able to call azd get-access-token but azd returned an error?  (for example, `azd not loged in`).  We should not move on to the next credential in the chain if azd is found but it is not loged in, should we?

Reply:
> If yes, should we return other error when we are able to call azd get-access-token but azd returned an error? (for example, azd not loged in). We should not move on to the next credential in the chain if azd is found but it is not loged in, should we?

I think that we should? DAC, for example, will move on in cases where `az` is installed but you are not logged in.

Here's the code for the chained credential in the JavaScript SDK:

https://github.com/Azure/azure-sdk-for-js/blob/3c3e424d0d259384a11318740ba250ab80586e37/sdk/identity/identity/src/credentials/chainedTokenCredential.ts#L61-L80

You can see that it uses both `CredentialUnavailableError` and `AuthenticationRequiredError` to move to the next credential in the chain.

So I think in the case where you are logged out, I should be throwing something like AuthenticationRequiredError, but we still want the move down the chain behavior.

Reply:
Would it make sense to support an environment variable like `LOCAL_DEBUG` ?
We would set this env var as part of `Start API` script and the azdCredential would automatically launch `azd login` if azd is not logged in. 
The user won't get an error in that case, but instead, azd would start the interactive flow

Reply:
I'm not really in love with the UX there and this would be a difference in behavior from the other tool credentials across the Azure SDK Ecosystem.  The message that comes back from the chained credential is good at surfacing errors from the individual credentials - let's start with just giving a good and simple error message there like: "no user logged in, use `azd login` to login" and see if that's good enough.